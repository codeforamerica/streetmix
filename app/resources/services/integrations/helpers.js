const { User } = require('../../../db/models')

/**
 finds the database record for the given user
 */
// might be better to do something besides return null
// we can actually probably get rid of this function if we want, but it might help
// keep some of the other code clean
const findUser = async function (userId) {
  const user = await User.findOne({ where: { auth0Id: userId } })
  if (user === null) {
    return null
  }
  return user.dataValues
}

// if the provider we're trying to find exists
// return its index value or -1 if not found
const addOrUpdateByProviderName = function (array, item) {
  const indexValue = array.findIndex(
    (_item) => _item.provider === item.provider
  )
  if (indexValue === -1) {
    array.push(item)
  } else {
    array[indexValue] = item
  }
}

/**
 given a user, check their identity data,
 depending on what data you find, add subscriber role

 TODO: make this less permissive(maybe add a status field or have logic depending on the provider), tested,
 and remove the role if we're confident we should (probably shouldn't worry about this for MVP)
 */
const syncAccountStatus = async function (userId) {
  const databaseUser = await User.findOne({ where: { auth0Id: userId } })
  if (databaseUser.identities.some((identity) => 'user_id' in identity)) {
    databaseUser.addRole('SUBSCRIBER_1')
  }
}

module.exports = { findUser, addOrUpdateByProviderName, syncAccountStatus }
