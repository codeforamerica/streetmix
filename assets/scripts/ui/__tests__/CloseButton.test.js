/* eslint-env jest */
import React from 'react'
import { fireEvent, cleanup } from '@testing-library/react'
import { renderWithIntl as render } from '../../../../test/helpers/render'
import CloseButton from '../CloseButton'

describe('CloseButton', () => {
  afterEach(cleanup)

  it('renders snapshot', () => {
    const wrapper = render(<CloseButton onClick={jest.fn()} />)
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('renders with custom title, class name, and other attributes', () => {
    const wrapper = render(<CloseButton
      onClick={jest.fn()}
      title="foofoo"
      className="my-class"
      disabled
      hidden
    />)
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should call onClick function when button is clicked', () => {
    const onClick = jest.fn()
    const { getByTitle } = render(<CloseButton
      onClick={onClick}
      title="foo"
    />)
    fireEvent.click(getByTitle('foo'))
    expect(onClick).toHaveBeenCalledTimes(1)
  })
})
