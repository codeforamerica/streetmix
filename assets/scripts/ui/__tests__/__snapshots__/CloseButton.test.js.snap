// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloseButton renders snapshot 1`] = `
<DocumentFragment>
  <button
    class="close"
    title="Dismiss"
  >
    <svg
      class="svg-inline--fa fa-times"
    />
  </button>
</DocumentFragment>
`;

exports[`CloseButton renders with custom title, class name, and other attributes 1`] = `
<DocumentFragment>
  <button
    class="my-class"
    disabled=""
    hidden=""
    title="foofoo"
  >
    <svg
      class="svg-inline--fa fa-times"
    />
  </button>
</DocumentFragment>
`;
